### ISSUE TITLE
Put $x/bin before $x/sbin in --user sessions' PATH

### ISSUE THREAD

**Issue by mathstuf** [systemd/systemd#733]
I've converted my user session to using `systemd --user`, but now I'm getting `/usr/sbin` before `/usr/bin` in the path. This causes problems with `mock` which has `/usr/sbin/mock` and `/usr/bin/mock` and the latter is the one meant to be run by a user.

This occurs within `tmux` from the shell because `tmux` was started with a user-level `tmux.service`. Also occurs in `urxvt` because I'm use the server/client mode via a `urxvt256cd.service`.

Would it be possible to place `/usr/bin` before `/usr/sbin` (and similar for `/usr/local`) for `--user` service files?

See https://bugzilla.redhat.com/show_bug.cgi?id=1246810

**Comment by poettering** [comment]
Having two binaries by the same name in bin and sbin sounds really sick, and should be fixed. I am not sure we want to support such a thing in the built-in defaults. Generally we give the more privileged, more basic tools precedence over the less privileged, more additional tools. This means "/sbin" gets precedence over "/bin" and "/usr/bin", if all of these exist.

You should be able to configure this locally, by editing ~/.config/systemd/user.conf and setting DefaultEnvironment= there.

**Comment by mathstuf** [comment]
I don't see in the documentation the defaults for the environment. Are they mentioned anywhere? If I set it, is it supplemental to systemd's internal value or does it replace it? Somewhat related: can I delete environment variables using it?

**Comment by poettering** [comment]
@mathstuf they aren't document, and they are different on the various distros (due to some having done the /usr-merge and others not having done it). I am not sure we should document this, since this isn't stable.

IIRC you can override the whole setting. Try it out.

**Comment by mathstuf** [comment]
Restarting the user session is really invasive due to #735 :) . I'll try to test it out tonight.

Maybe at least the behavior in relation to the defaults should be specified in the docs though?

**Comment by poettering** [comment]
yeah, we should document how $PATH is set. We currently mention it in systemd.exec in the "Environment Variables in Spawned Processes" part, but we should probably mention there that user.conf and systemd.conf can be used to alter it.

**Comment by keszybz** [comment]
On second thought, putting /bin before /sbin is user $PATH is fairly standard. Maybe we should flip the two for user session... It's not going to break anything for people with "sane" setups where they only have one binary with the given name in $PATH.

**Comment by poettering** [comment]
Naahh, we shouls tick to one order and one order only. Having different order for user and system is just wrong...

If you say "fairly standard", what do you mean precisely? Can you link to examples for the order being used?

**Comment by mathstuf** [comment]
Right from `/etc/profile`:

``` sh
# Path manipulation
if [ "$EUID" = "0" ]; then
    pathmunge /usr/sbin
    pathmunge /usr/local/sbin
else
    pathmunge /usr/local/sbin after
    pathmunge /usr/sbin after
fi
```

**Comment by keszybz** [comment]
The above is from Fedora. Debian has something similar:

```
ENV_SUPATH      PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
ENV_PATH        PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
```

login(1) says:

> $PATH defaults  to  /usr/local/bin:/bin:/usr/bin  for normal users, and to /usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin for root, if not otherwise configured.

Running strings on /bin/login confirms that the man page matches the binary. But in Fedora /bin and /sbin/ are at the end, so this setting is overridden somewhere (not in /etc/profile, because that would result in a /usr/sbin being after /sbin, but it's before). But I cannot find the place.

**Comment by keszybz** [comment]
I did some more digging:
sshd uses: `/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin` and `/usr/local/bin:/usr/bin` by default.

gdm uses `/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin`. I filed https://bugzilla.redhat.com/show_bug.cgi?id=1251192 to have the last two component trimmed.

**Comment by halfline** [comment]
bin should always be before sbin in the user session path.  consolehelper and other legacy tools rely on it.

( i mean the whole premise of consolehelper is bad, but it's been around forever, and we shouldn't break it unintentionally without a real reason )

**Comment by keszybz** [comment]
> we should probably mention there that user.conf and systemd.conf can be used to alter it.

I started to add this, but it is already documented at bottom of that section, and a link to systemd-system.conf(5) is provided.

**Comment by prajnoha** [comment]
Fixed with commit 2e90c6f5af0a3a9149f3e44f7b7038d11bb9b12a
